### PR TITLE
backport: feat(ui): add sync status panel to wallet screen

### PR DESCRIPTION
**Pull request by lklimek**
## Summary

- Adds a compact sync status panel between wallet selection and wallet detail panel
- Core line shows: RPC connected/disconnected, or SPV with phase-specific progress (Headers/Filter Headers/Filters/Blocks with %)
- Platform line shows: address count, sync block height, and relative "time ago" since last sync
- Adds `format_unix_time_ago()` / `format_duration_ago()` time formatting helpers

## Test plan

Manual test scenarios: [`docs/ai-design/2026-02-24-sync-status-panel/manual-test-scenarios.md`](docs/ai-design/2026-02-24-sync-status-panel/manual-test-scenarios.md)

- [ ] Panel visible when HD wallet selected, hidden otherwise
- [ ] RPC mode: shows Connected/Disconnected
- [ ] SPV mode: shows sync phase with progress percentage
- [ ] Platform: shows address count, height, time-ago
- [ ] Wallet switching updates panel
- [ ] Light/dark mode theming

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

## Summary by CodeRabbit

* **New Features**
  * Sync Status Panel on Wallets screen (Core RPC/SPV + per-wallet Platform sync) with SPV progress text and time-ago timestamps (developer mode).
  * Connection status tooltip now shows richer SPV progress and unbanned-vs-total endpoint info.

* **Bug Fixes**
  * Prevent stale in-memory transaction tracking on asset lock broadcast failures.
  * Replace crash-prone unwraps in signing flows with graceful error handling.

* **Documentation**
  * Added comprehensive manual test scenarios for the Sync Status Panel.

### CODE REVIEW

**Comment by lklimek** [comment]
In SPV mode, I keep having `Platform: Addresses: never synced`. We need to investigate that.

**Comment by PastaPastaPasta** [comment]
Screenshot 2026-02-24 at 19 25 17
Screenshot 2026-02-24 at 19 25 35
Screenshot 2026-02-24 at 19 27 55


Concept NACK. I don't think it makes sense to show this on the wallets page; this is relevant to most DET operations.

Maybe instead we enhance the sync indicator in top left to also show progress, but I don't like this approach.

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
concept NACK

**Comment by lklimek** [comment]
## 🔍 Audit Summary — PR #642

**Reviewed by:** Claude Code with a 3-agent team:
- `rust-developer` — Code quality, idiomatic patterns, error handling, lock safety
- `security-engineer` — Security audit, OWASP classification, state consistency
- `project-reviewer` — Convention adherence, cross-file consistency, test coverage

**Overall risk: Low.** No critical or high-severity issues. The PR improves codebase robustness (panic removal, Mutex error handling) while adding a well-gated diagnostic UI panel.

---

### Findings

| # | Severity | Location | Description |
|---|----------|----------|-------------|
| 1 | 🟡 Medium | `create_asset_lock.rs:43-56` | **UTXO state inconsistency on broadcast failure** — Comment says "we trigger a wallet refresh" but code does not. UTXOs removed from memory but never spent on-chain → incorrect balance until manual refresh. [A04:2021] |
| 2 | 🟡 Medium | `wallets_screen/mod.rs:1305-1347` | **Missing masternodes sync phase + duplicated SPV phase logic** — `spv_active_phase_text` omits the Masternodes phase (present in `NetworkChooserScreen::format_sync_progress`) and duplicates phase interpretation logic. Users see generic "syncing..." during masternodes phase. |
| 3 | 🟡 Medium | `wallets_screen/mod.rs:820-843` | **Utility functions trapped on screen struct** — `format_unix_time_ago`/`format_duration_ago` are pure functions with no `self` dependency. Should live in a shared utility module for reuse. |
| 4 | 🟡 Medium | `create_asset_lock.rs:52` | **Silent Mutex poisoning in cleanup path** — Happy path propagates lock errors, but cleanup uses `if let Ok(...)` which silently ignores poisoned Mutex. Should at minimum log a warning. [A04:2021] |
| 5 | 🟢 Low | `wallets_screen/mod.rs` | **No unit tests for new helper functions** — `format_duration_ago`, `simple_progress_pct`, `spv_active_phase_text` are pure logic, trivially testable, but have no tests. |
| 6 | 🟢 Low | `manual-test-scenarios.md` | **TS-25 references work not in this PR** — Describes `shared_connection` removal, but `Database::shared_connection()` still exists and the diff doesn't touch it. |

### Pre-existing / Outside-diff observations

| # | Severity | Description |
|---|----------|-------------|
| P1 | 🟡 Medium | **11 remaining `.unwrap()` on `transactions_waiting_for_finality` Mutex** across `transaction_processing.rs`, `register_identity.rs`, `top_up_identity.rs`, `fund_platform_address_from_wallet_utxos.rs`. Any panic there poisons the Mutex for the new graceful callers. Track as follow-up. |

### ✅ Positive observations

- **Panic removal in signing code** (`asset_lock_transaction.rs`) — `.expect()` → proper `Result` propagation with descriptive error messages. Excellent hardening.
- **Broadcast failure cleanup** — Removing stale `transactions_waiting_for_finality` entries on failure prevents phantom "waiting for finality" states.
- **Dev-mode gating** — Sync panel behind `is_developer_mode()` + collapsible. Good UX design.
- **`i64` → `i128` cast** for fee diff logging prevents overflow on large credit values.
- **`saturating_sub` for time calculations** — Prevents underflow on clock skew.
- **Division-by-zero guard** in `simple_progress_pct` with `target == 0` check.
- **Connection status tooltip improvement** — "5 unbanned / 10 total" is clearer than "5/10".
- **Manual test scenarios** are thorough (24 cases + edge cases table) and correctly placed per project conventions.

### Redundancy analysis

3 agents produced 18 raw findings. After deduplication: **6 unique actionable findings + 1 pre-existing observation**. Redundancy ratio: 39% (7 findings flagged by 2+ agents). Key overlap areas: Mutex handling strategy, SPV phase logic concerns, and time formatting edge cases.

---

<sub>🤖 Reviewed by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent | `rust-developer` · `security-engineer` · `project-reviewer`</sub>

File: src/backend_task/core/create_asset_lock.rs [hunk 1]
```diff
@@ -33,14 +33,27 @@ impl AppContext {
...
+                .map_err(|e| e.to_string())?;
             proofs.insert(tx_id, None);
         }
 
-        // Broadcast the transaction
-        self.broadcast_raw_transaction(&asset_lock_transaction)
+        // Broadcast the transaction.  If broadcast fails, the UTXOs have already
+        // been removed from the wallet (inside the transaction builder) but were
+        // never actually spent on-chain.  We trigger a wallet refresh so the next
+        // UTXO reload reconciles the in-memory state with the chain.
```
Comment (lklimek) [inline src/backend_task/core/create_asset_lock.rs]: 🟡 **Medium** [A04:2021-Insecure Design] — **Misleading comment: no wallet refresh actually triggered**

The comment says "We trigger a wallet refresh so the next UTXO reload reconciles the in-memory state with the chain" — but the code below does NOT trigger any refresh. It only cleans up the finality tracking entry and returns an error.

After a failed broadcast, UTXOs have already been removed from the wallet's in-memory set (inside `remove_selected_utxos`) but were never spent on-chain. The user sees an incorrect lower balance until manual refresh.

Options:
1. Actually trigger a wallet refresh in the error path (return a result variant that tells the UI to refresh)
2. Restore the removed UTXOs in the error path
3. At minimum, fix the comment to accurately describe current behavior and add a `TODO`

File: src/backend_task/core/create_asset_lock.rs [hunk 2]
```diff
@@ -33,14 +33,27 @@ impl AppContext {
...
+        if let Err(e) = self
+            .broadcast_raw_transaction(&asset_lock_transaction)
             .await
-            .map_err(|e| format!("Failed to broadcast asset lock transaction: {}", e))?;
+        {
+            // Clean up the finality tracking entry
+            if let Ok(mut proofs) = self.transactions_waiting_for_finality.lock() {
```
Comment (lklimek) [inline src/backend_task/core/create_asset_lock.rs:53]: 🟡 **Medium** — **Silent Mutex poisoning in cleanup path**

The happy path (line 36) correctly propagates poisoning via `map_err`, but this cleanup path silently ignores it. A poisoned Mutex indicates a prior panic — a stale entry in `transactions_waiting_for_finality` could cause the app to wait indefinitely for finality of a never-broadcast tx.

```suggestion
            if let Ok(mut proofs) = self.transactions_waiting_for_finality.lock() {
                proofs.remove(&tx_id);
            } else {
                tracing::warn!("Failed to clean up finality tracking for tx {tx_id}: Mutex poisoned");
            }
```

File: src/ui/wallets/wallets_screen/mod.rs [hunk 3]
```diff
@@ -1105,6 +1151,208 @@ impl WalletsBalancesScreen {
...
+                                },
+                            ));
+                        });
+                    });
+            },
+        );
+    }
+
+    /// Get a text summary of the active SPV sync phase.
+    fn spv_active_phase_text(sync_progress: &Option<SpvSyncProgress>) -> String {
```
Comment (lklimek) [inline src/ui/wallets/wallets_screen/mod.rs]: 🟡 **Medium** — **Duplicated SPV phase logic + missing Masternodes phase**

`NetworkChooserScreen::format_sync_progress` (network_chooser_screen.rs ~L1961) already interprets `SpvSyncProgress` phases and includes a Masternodes phase that this function omits. During a fresh sync where masternodes is the active phase, users see generic "syncing..." instead of a meaningful label.

Consider:
1. Extract shared SPV phase interpretation into a utility (e.g., `src/ui/spv_helpers.rs`) that both screens call
2. Or at minimum, add masternodes handling here and document the intentional simplification

File: src/ui/wallets/wallets_screen/mod.rs [hunk 4]
```diff
@@ -797,6 +820,29 @@ impl WalletsBalancesScreen {
         Amount::dash_from_duffs(amount_duffs).to_string()
```
Comment (lklimek) [inline src/ui/wallets/wallets_screen/mod.rs:833]: 🟡 **Medium** — **Pure utility functions trapped on a screen struct**

`format_unix_time_ago` and `format_duration_ago` have no dependency on `self` or `WalletsBalancesScreen`. Time-ago formatting is a general need that will likely surface on other screens (identity last-seen, contest deadlines, etc.).

Consider moving these to a shared utility module (e.g., `src/ui/helpers.rs` or `src/utils/time.rs`) as free functions.

File: src/ui/wallets/wallets_screen/mod.rs [hunk 5]
```diff
@@ -1105,6 +1151,208 @@ impl WalletsBalancesScreen {
...
+        }
+
+        if let Ok(blocks) = progress.blocks()
+            && blocks.state() == SyncState::Syncing
+        {
+            let pct = Self::simple_progress_pct(blocks.last_processed(), target_height);
+            return format!("Blocks {pct}%");
+        }
+
+        "syncing...".to_string()
```
Comment (lklimek) [inline src/ui/wallets/wallets_screen/mod.rs]: 🟢 **Low** — **Fallback text "syncing..." when no phase is active**

If all four phases return non-`Syncing` states (or `Err`), this fallback fires. Worth noting that `target_height` (line 1311-1315) silently becomes 0 if `progress.headers()` returns `Err`, which would make the Blocks branch show "Blocks 0%" even during active sync.

Also, this function has no unit tests despite being pure logic — the boundary conditions (all phases done, headers Err, target=0) are good test candidates.

File: docs/ai-design/2026-02-24-sync-status-panel/manual-test-scenarios.md [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+## TS-25: Database refactor -- shared_connection removal
+
+### Preconditions
+- This is a code-level verification, not a UI test.
+
+### Steps
+1. Verify the application compiles without errors (`cargo build`).
+2. Run the full test suite (`cargo test --all-features --workspace`).
+3. Search the codebase for any remaining calls to `Database::shared_connection()`.
+
```
Comment (lklimek) [inline docs/ai-design/2026-02-24-sync-status-panel/manual-test-scenarios.md]: 🟢 **Low** — **TS-25 references work not in this PR**

This scenario describes verification of `shared_connection()` removal, but `Database::shared_connection()` still exists at `src/database/mod.rs:60` and the diff doesn't touch that file. This test scenario appears to belong to a different PR.

Consider removing TS-25 from this document to avoid confusion.

File: src/ui/wallets/wallets_screen/mod.rs [hunk 7]
```diff
@@ -173,6 +176,13 @@ impl WalletsBalancesScreen {
         selected_wallet: Option<Arc<RwLock<Wallet>>>,
         selected_single_key_wallet: Option<Arc<RwLock<SingleKeyWallet>>>,
     ) -> Self {
+        let platform_sync_info = selected_wallet
+            .as_ref()
+            .and_then(|w| w.read().ok().map(|g| g.seed_hash()))
+            .and_then(|hash| app_context.db.get_platform_sync_info(&hash).ok())
+            .map(|(ts, checkpoint, _terminal)| (ts, checkpoint))
+            .filter(|(ts, _)| *ts > 0);
```
[the remark this answers is not in the document]
Comment (lklimek) [reply src/ui/wallets/wallets_screen/mod.rs:184]: Won't fix — the `ts > 0` filter is intentional. When `ts == 0`, the wallet has genuinely never completed a platform sync (the DB stores 0 as the default). Showing "never synced" is correct in this case. A checkpoint without a timestamp would indicate incomplete/corrupt data that we don't want to display as valid sync state.

File: src/ui/wallets/wallets_screen/mod.rs (hunk 4 above)
Comment (lklimek) [reply src/ui/wallets/wallets_screen/mod.rs:833]: Acknowledged — deferring to a follow-up. These are the only callers right now and extracting a utility module is out of scope for this PR.

File: src/ui/wallets/wallets_screen/mod.rs (hunk 5 above)
Comment (lklimek) [reply src/ui/wallets/wallets_screen/mod.rs]: Acknowledged — the dev-mode panel is tested via manual test scenarios (TS-05 through TS-24). Unit tests for helpers can be added in a follow-up.

File: src/ui/wallets/wallets_screen/mod.rs [hunk 8]
```diff
@@ -116,6 +117,8 @@ pub struct WalletsBalancesScreen {
     utxo_page: usize,
     /// Selected refresh mode (only shown in dev mode)
     refresh_mode: RefreshMode,
+    /// Cached platform sync info: (last_sync_timestamp, last_sync_height)
+    platform_sync_info: Option<(u64, u64)>,
```
[the remark this answers is not in the document]
Comment (lklimek) [reply src/ui/wallets/wallets_screen/mod.rs:121]: Fixed — added `set_selected_hd_wallet()` helper that centralizes wallet selection + cache refresh. All direct assignments to `self.selected_wallet` now go through this helper (or explicitly clear `platform_sync_info` for single-key wallet paths). Tracked in #657 for any remaining edge cases.

File: src/backend_task/core/create_asset_lock.rs (hunk 1 above)
Comment (lklimek) [reply src/backend_task/core/create_asset_lock.rs]: Fixed — updated the comment to accurately describe current behavior and added a reference to #657 for the proper wallet refresh follow-up.

File: src/backend_task/core/create_asset_lock.rs (hunk 2 above)
Comment (lklimek) [reply src/backend_task/core/create_asset_lock.rs:53]: Fixed — added `tracing::warn!` in both registration and top-up cleanup paths when the Mutex is poisoned.

File: docs/ai-design/2026-02-24-sync-status-panel/manual-test-scenarios.md (hunk 6 above)
Comment (lklimek) [reply docs/ai-design/2026-02-24-sync-status-panel/manual-test-scenarios.md]: Fixed — removed TS-25 from this PR's test scenarios document.
